actors/migration/miner_test.go

```diff
@@ -0,0 +1,492 @@
+	msectors, err := miner0.LoadSectors(store, st.Sectors)
+	require.NoError(t, err)
+	targetDeadline := miner0.NewDeadlineInfo(st.ProvingPeriodStart, dlIdx, rescheduleAndFaultEpoch).NextNotElapsed()
+	faultExpiration := targetDeadline.Last() + miner0.FaultMaxAge
+	require.True(t, faultExpiration < expectedExpiration)
+	partitionSectors := make(miner0.PartitionSectorMap)
+	require.NoError(t, partitionSectors.AddValues(pIdx, uint64(sectorNo)))
+	newFaultyPower, err := deadline.DeclareFaults(store, msectors, info.SectorSize, quant, faultExpiration, partitionSectors)
+	require.NoError(t, err)
+
+	require.NoError(t, deadlines.UpdateDeadline(store, dlIdx, deadline))
+	require.NoError(t, st.SaveDeadlines(store, deadlines))
+
+	require.NoError(t, pst.AddToClaim(store, maddr, newFaultyPower.Raw.Neg(), newFaultyPower.QA.Neg()))
+
+	// Show the state is b0rked
+	_, deadline, partition = loadDeadlineAndPartition0(t, st, store, dlIdx, pIdx)
+	advancedExpiration := quant.QuantizeUp(rescheduleAndFaultEpoch)
+	require.True(t, advancedExpiration < faultExpiration)
+	assertDeadlineExpirationQueue0(t, store, deadline, pIdx, map[abi.ChainEpoch][]uint64{
+		advancedExpiration: {pIdx}, // Appearing repeatedly here is ok, this queue is best-effort
+		faultExpiration:    {pIdx},
+		expectedExpiration: {pIdx},
+	})
+	assertPartitionExpirationQueue0(t, store, partition, map[abi.ChainEpoch]*miner0.ExpirationSet{
+		// The sector should be scheduled here, but with faulty power.
+		// It's wrongly here with active power, because fault declaration didn't find it.
+		advancedExpiration: {
+			OnTimeSectors: bitfield.NewFromSet([]uint64{uint64(sectorNo)}),
+			EarlySectors:  bitfield.New(),
+			OnTimePledge:  big.Zero(),
+			ActivePower:   oneSectorPower, // Power should be faulty
+			FaultyPower:   miner0.NewPowerPairZero(),
+		},
+		// The sector should not be scheduled here at all.
+		// It's wrongly here because fault declaration didn't find it at advancedExpiration.
+		faultExpiration: {
+			OnTimeSectors: bitfield.New(),
+			EarlySectors:  bitfield.NewFromSet([]uint64{uint64(sectorNo)}),
+			OnTimePledge:  big.Zero(),
+			ActivePower:   miner0.NewPowerPairZero(),
+			FaultyPower:   oneSectorPower,
+		},
+		// The non-rescheduled sector is still here.
+		// This wrongly has lost power twice.
+		expectedExpiration: {
+			OnTimeSectors: bitfield.NewFromSet([]uint64{uint64(otherSectorNo)}),
+			EarlySectors:  bitfield.New(),
+			OnTimePledge:  big.Zero(),
+			ActivePower:   miner0.NewPowerPairZero(), // Should have oneSectorPower for the other sector
+			FaultyPower:   miner0.NewPowerPairZero(),
+		},
+	})
+
+	phead, err = store.Put(ctx, pst)
+	require.NoError(t, err)
+	setActorState(t, tree, builtin0.StoragePowerActorAddr, builtin0.StoragePowerActorCodeID, phead)
+
+	mhead, err = store.Put(ctx, st)
+	require.NoError(t, err)
+	setActorState(t, tree, maddr, builtin0.StorageMinerActorCodeID, mhead)
```

Comment:
It would be a nice cross-check of `CheckStateInvariants` to run it here and assert that it's finding the problem.

Reply:
I guess that's only possible if we back port it to v0.9.